app/models/thredded/user_topic_follow.rb

```diff
@@ -13,8 +13,19 @@ class UserTopicFollow < ActiveRecord::Base
     alias_attribute :postable_id, :topic_id
     alias_attribute :postable, :topic
 
-    def self.create_unless_exists!(user_id, topic_id, reason = :manual)
-      create_with(reason: reason).find_or_create_by!(user_id: user_id, topic_id: topic_id)
+    # Creates a follow or finds the existing one.
+    #
+    # This method is safe to call concurrently from different processes. Lookup and creation happen in a transaction.
+    # If an ActiveRecord::RecordNotUnique error is raised, the find is retried.
+    #
+    # @return [Thredded::UserTopicFollow]
+    def self.create_unless_exists(user_id, topic_id, reason = :manual)
+      transaction(requires_new: true) do
+        create_with(reason: reason).find_or_create_by(user_id: user_id, topic_id: topic_id)
+      end
+    rescue ActiveRecord::RecordNotUnique
+      # The record has been created from another connection, retry to find it.
+      retry
```

Comment:
I don't quite get this one... 

I get the fact that `find_or_create_by` is not necessarily a single action, so db change could happen between them and then this could fail. And if I understand correctly, the new transaction is to ensure that the isolation happens immediately (rather than being deferred to a wider wrapping transaction which could mean the whole thing fails for no reason).

But if so, then there is no need for a retry -- the retry will only successfully find the record.

Or are there circumstances when ActiveRecord:RecordNotUnique may be raised and the row is not reliably (outside a transaction) in the database?